### PR TITLE
API for immutable Secrets and ConfigMaps

### PR DESCRIPTION
**Pull request by wojtek-t**
This implements the API part of the KEP (allow for marking secrets/configmaps as immutable + necessary bits to enforce that and e2e tests for it).

The Kubelet part is not yet implemented (those this part itself solve the first user problem being ability to protect from accidental bad pushes).

Part of https://github.com/kubernetes/enhancements/issues/1412

```release-note
Introduce Alpha field `Immutable` in both Secret and ConfigMap objects to mark their contents as immutable. The implementation is hidden behind feature gate `ImmutableEphemeralVolumes` (currently in Alpha stage).
```

### CODE REVIEW

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -5354,6 +5354,10 @@
           "description": "Data contains the configuration data. Each key must consist of alphanumeric characters, '-', '_' or '.'. Values with non-UTF-8 byte sequences must use the BinaryData field. The keys stored in Data must not overlap with the keys in the BinaryData field, this is enforced during validation process.",
           "type": "object"
         },
+        "immutable": {
+          "description": "Immutable field, if set, ensures that data stored in the configmap cannot be updated (only object metadata can be modified). This is an alpha field enabled by ImmutableEphemeralVolumes feature gate.",
```
Comment (sftim) [inline api/openapi-spec/swagger.json]: nit, capitalization
```suggestion
          "description": "Immutable field, if set, ensures that data stored in the ConfigMap cannot be updated (only object metadata can be modified). This is an alpha field enabled by ImmutableEphemeralVolumes feature gate.",
```

Comment (wojtek-t) [reply api/openapi-spec/swagger.json]: Done

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -5051,8 +5061,20 @@ func ValidateConfigMap(cfg *core.ConfigMap) field.ErrorList {
 func ValidateConfigMapUpdate(newCfg, oldCfg *core.ConfigMap) field.ErrorList {
 	allErrs := field.ErrorList{}
 	allErrs = append(allErrs, ValidateObjectMetaUpdate(&newCfg.ObjectMeta, &oldCfg.ObjectMeta, field.NewPath("metadata"))...)
-	allErrs = append(allErrs, ValidateConfigMap(newCfg)...)
 
+	if oldCfg.Immutable != nil && *oldCfg.Immutable {
```
Comment (saad-ali) [inline pkg/apis/core/validation/validation.go]: Does this check also need to be behind feature gate?

File: pkg/registry/core/configmap/strategy.go [hunk 3]
```diff
@@ -86,6 +86,21 @@ func (strategy) ValidateUpdate(ctx context.Context, newObj, oldObj runtime.Objec
 	return validation.ValidateConfigMapUpdate(newCfg, oldCfg)
 }
 
+func dropDisabledFields(configMap *api.ConfigMap, oldConfigMap *api.ConfigMap) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ImmutableEphemeralVolumes) && oldConfigMap == nil {
+		configMap.Immutable = nil
```
Comment (saad-ali) [inline pkg/registry/core/configmap/strategy.go]: Will this result in silently dropping the new `Immutable` field if feature gate is disabled? If so, how about explicitly failing if field is set but feature is disabled?

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -5005,6 +5005,16 @@ func ValidateSecretUpdate(newSecret, oldSecret *core.Secret) field.ErrorList {
 	}
 
 	allErrs = append(allErrs, ValidateImmutableField(newSecret.Type, oldSecret.Type, field.NewPath("type"))...)
+	if oldSecret.Immutable != nil && *oldSecret.Immutable {
```
Comment (saad-ali) [inline pkg/apis/core/validation/validation.go]: Please add unit tests for new behavior.

**Review by saad-ali — COMMENTED** [review]
> The Kubelet part is not yet implemented (those this part itself solve the first user problem being ability to protect from accidental bad pushes).

Is the plan to get the kubelet changes in this release as well?

File: pkg/registry/core/configmap/strategy.go (hunk 3 above)
Comment (wojtek-t) [reply pkg/registry/core/configmap/strategy.go]: This is the pattern we use for all newly added fields, so I think it's better to be consistent.

**Comment by wojtek-t** [comment]
> Is the plan to get the kubelet changes in this release as well?

Yes - I have them as part of my POC - just need to clean them up.
That said, I think this PR is self-contained, so I was thinking about doing them as a separate PR.

Re other comments - thanks, will address them.

File: pkg/apis/core/validation/validation.go (hunk 4 above)
Comment (wojtek-t) [reply pkg/apis/core/validation/validation.go]: Done (both here and for config map)

File: pkg/apis/core/validation/validation.go (hunk 2 above)
Comment (wojtek-t) [reply pkg/apis/core/validation/validation.go]: My understanding is it doesn't have to.
The rationale is:
- if the feature gate is off, then the "dropDisabledFields" logic from PrepareForUpdate is firing
- this is happening before validation, so the Immutable field will be clear in this case

**Review by wojtek-t — COMMENTED** [review]
@saad-ali - thanks for the review.
Comments addressed or responded to. PTAL

**Comment by wojtek-t** [comment]
Alpha tests failing due to: https://github.com/kubernetes/kubernetes/issues/86908 - seems unrelated to this PR.

File: pkg/apis/core/validation/validation.go [hunk 5]
```diff
@@ -5005,6 +5005,16 @@ func ValidateSecretUpdate(newSecret, oldSecret *core.Secret) field.ErrorList {
...
+		if !reflect.DeepEqual(newSecret.Immutable, oldSecret.Immutable) {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("immutable"), "field is immutable when immutable is set"))
```
Comment (tedyu) [inline pkg/apis/core/validation/validation.go]: It seems 'when immutable is set' is redundant

File: pkg/registry/core/secret/strategy.go [hunk 6]
```diff
@@ -67,12 +71,26 @@ func (strategy) AllowCreateOnUpdate() bool {
...
+	oldSecret := old.(*api.Secret)
+	dropDisabledFields(newSecret, oldSecret)
 }
 
 func (strategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
 	return validation.ValidateSecretUpdate(obj.(*api.Secret), old.(*api.Secret))
 }
 
+func dropDisabledFields(secret *api.Secret, oldSecret *api.Secret) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ImmutableEphemeralVolumes) && oldSecret == nil {
```
Comment (tedyu) [inline pkg/registry/core/secret/strategy.go]: It seems we only need to call DefaultFeatureGate.Enabled once in this func

File: pkg/registry/core/configmap/strategy.go (hunk 3 above)
Comment (saad-ali) [reply pkg/registry/core/configmap/strategy.go]: Ack, will leave it to API reviewers to verify

File: pkg/apis/core/validation/validation.go (hunk 2 above)
Comment (saad-ali) [reply pkg/apis/core/validation/validation.go]: SGTM

**Review by saad-ali — APPROVED** [review]
/lgtm

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (wojtek-t) [reply pkg/apis/core/validation/validation.go]: I don't fully agree - if "immutable" is unset, this field is actually mutable.

File: pkg/registry/core/secret/strategy.go (hunk 6 above)
Comment (wojtek-t) [reply pkg/registry/core/secret/strategy.go]: Technically probably true, but I think this one is easier to reason about in my opinion (for insignificant performance drop).

**Comment by wojtek-t** [comment]
Thanks @saad-ali !

**Comment by wojtek-t** [comment]
/assign @thockin 
For API approval (already reviewed by Saad (api-reviewer)).

File: staging/src/k8s.io/api/core/v1/types.go [hunk 7]
```diff
@@ -5424,6 +5424,12 @@ type Secret struct {
 	// +optional
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
+	// Immutable field, if set, ensures that data stored in the Secret cannot
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go]: Since comments become docs...

"Immutable ensures that data ..." (basically drop 'field')

Comment (thockin) [reply staging/src/k8s.io/api/core/v1/types.go]: Document the default?

Comment (thockin) [reply staging/src/k8s.io/api/core/v1/types.go]: Is this field updatable or only set at creation?

Comment (thockin) [reply staging/src/k8s.io/api/core/v1/types.go]: Can this field change from false to true?

File: pkg/registry/core/secret/strategy.go (hunk 6 above)
Comment (thockin) [reply pkg/registry/core/secret/strategy.go]: Most times I see this call an "is in use" funtion that encapsulates both old == nil and old.field == nil.

File: pkg/registry/core/configmap/strategy.go (hunk 3 above)
Comment (thockin) [reply pkg/registry/core/configmap/strategy.go]: If teh gate is disabled we want to pretend the field did not exist.  We do not error on unknown fields

File: pkg/apis/core/validation/validation.go [hunk 8]
```diff
@@ -5005,6 +5005,16 @@ func ValidateSecretUpdate(newSecret, oldSecret *core.Secret) field.ErrorList {
 	}
 
 	allErrs = append(allErrs, ValidateImmutableField(newSecret.Type, oldSecret.Type, field.NewPath("type"))...)
+	if oldSecret.Immutable != nil && *oldSecret.Immutable {
+		if !reflect.DeepEqual(newSecret.Immutable, oldSecret.Immutable) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: Don't use reflect for this - it's just a bool right?  Whan I see reflect, I assume something tricky and subtle is happening...is it?

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go]: Agree with @wojtek-t 

field names in errors should be back-quoted: "when `immutable` is set"

File: staging/src/k8s.io/api/core/v1/types.go (hunk 7 above)
Comment (wojtek-t) [reply staging/src/k8s.io/api/core/v1/types.go]: Addressed all (here and in configmap).

**Review by wojtek-t — COMMENTED** [review]
@thockin - comments addressed; PTAL

**Comment by wojtek-t** [comment]
@thockin - this is now ready for the second pass (test failures are unrelated flakes)

/retest

File: pkg/apis/core/validation/validation.go (hunk 8 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go]: yes, IMO

**Review by thockin — COMMENTED** [review]
Last bit can be a followup

Thanks!

/lgtm
/approve

File: pkg/apis/core/validation/validation.go (hunk 8 above)
Comment (wojtek-t) [reply pkg/apis/core/validation/validation.go]: Opened https://github.com/kubernetes/kubernetes/pull/87602 for the cleanup

**Pull request merged**